lib/puppet/parser/resource.rb

```diff
@@ -94,8 +94,18 @@ def exported=(value)
     end
   end
 
-  # Do any finishing work on this object, called before evaluation or
-  # before storage/translation. The method does nothing the second time
+  # Finish the evaluation by assigning defaults and scope tags
+  # @api private
+  #
+  def finish_evaluation
+    return if @evaluation_finished
+    add_defaults
+    add_scope_tags
+    @evaluation_finished = true
```

Comment:
`finish` sets its guard flag early - presumably to make sure it gets set even in the case of an exception. Should the same thing be done here?